packages/protocol/contracts/dapis/SamplePriceDataFeed.sol

```diff
@@ -19,6 +37,170 @@ contract SamplePriceDataFeed {
+    bytes32 _templateId,
+    address _designatedWallet
+  ) external onlyAdminOrMetaAdmin cooldownTimeEnded {
+    // PriceDataFeed fetches the current dAPI parameters from RrpDapiServer
+    (
+      uint256 noResponsesToReduce,
+      uint256 toleranceInPercentages,
+      uint256 requesterIndex,
+      bytes32[] memory templateIds,
+      address[] memory designatedWallets,
+      address reduceAddress,
+      bytes4 reduceFunctionId,
+      address requestIndexResetter
+    ) = rrpDapiServer.getDapi(_dapiId);
+    // TODO: check dAPI was found?
+
+    // Adds templateId to dAPI.templateIds (iterating over because array push() is only available in storage arrays)
+    // Also add _designatedWallet to dAPI.designatedWallets
+    bytes32[] memory newTemplateIds = new bytes32[](templateIds.length + 1);
+    address[] memory newDesignatedWallets = new address[](designatedWallets.length + 1);
+    for (uint256 i = 0; i < templateIds.length; i++) {
+      newTemplateIds[i] = templateIds[i];
+      newDesignatedWallets[i] = designatedWallets[i];
+    }
+    newTemplateIds[newTemplateIds.length - 1] = _templateId;
+    newDesignatedWallets[newDesignatedWallets.length - 1] = _designatedWallet;
+
+    // Calls RrpDapiServer to create the new dAPI
+    bytes16 newDapiId =
+      rrpDapiServer.registerDapi(
+        noResponsesToReduce,
+        toleranceInPercentages,
+        requesterIndex,
+        newTemplateIds,
+        newDesignatedWallets,
+        reduceAddress,
+        reduceFunctionId,
+        requestIndexResetter
+      );
+
+    // Update dapiLastUpdated timestamp
+    adminToDapiLastUpdated[msg.sender] = uint64(block.timestamp);
+
+    // Record the returned dapiId
+    latestDapiId = newDapiId;
+
+    // TODO: publish TemplateAdded event?
+  }
+
+  function removeTemplate(bytes16 _dapiId, bytes32 _templateId) external onlyAdminOrMetaAdmin cooldownTimeEnded {
+    // PriceDataFeed fetches the current dAPI parameters from RrpDapiServer
+    (
+      uint256 noResponsesToReduce,
+      uint256 toleranceInPercentages,
+      uint256 requesterIndex,
+      bytes32[] memory templateIds,
+      address[] memory designatedWallets,
+      address reduceAddress,
+      bytes4 reduceFunctionId,
+      address requestIndexResetter
+    ) = rrpDapiServer.getDapi(_dapiId);
+    // TODO: check dAPI was found?
+
+    // Removes templateId from dAPI.templateIds (iterating over because array pop() is only available in storage arrays)
+    // Also removes designatedWallet from dAPI.designatedWallets
+    bool found = false;
+    bytes32[] memory newTemplateIds = new bytes32[](templateIds.length - 1);
+    address[] memory newDesignatedWallets = new address[](designatedWallets.length - 1);
+    for (uint256 i = 0; i < templateIds.length; i++) {
+      // copy array without _templateId
+      if (templateIds[i] == _templateId) {
+        found = true;
+      }
+      uint256 j = found ? i + 1 : i;
+      if (i != templateIds.length - 1) {
```

Comment:
This is at risk of being really gas cost inefficient. I think we might wanna consider a different data structure to prevent looping.

Reply:
Not really sure why but that's a good observation. I'll research more on why and how to improve this.

Reply:
Yea, this is very inefficient, although it's not your implementation, it's the dAPI data structure.

If we must be accessing individual templates in the dAPI struct, then it should be a mapping @bbenligiray

Reply:
Something kind of related is also the designatedWallets array. It seems that there's a 1 to 1 relationship between designated wallets and template ids and that makes me think that all this info might be better stored together using a struct.

Reply:
@sassmilic @bbenligiray I've been thinking and maybe the best way to avoid having to iterate over the templateIds/designatedWallets arrays is to have the addTemplate, updateTemplate and removeTemplate methods also in the RrpDapiServer contract. This way we can access the storage array directly and use push/pop methods. Maybe for update we'll still need to iterate tho.

Reply:
Even though we can add another mapping to store the index in the array for each templateId (`mapping(bytes32 => uint)`) I'm still struggling to completely get rid of the for loops since we must make a copy of both arrays (templateIds and designatedWallets) before calling `registerDapi()` function.

Reply:
Ahh, this is what I said this morning's meeting. I didn't realize you already mentioned it!

Reply:
Can you show me where in the code you're referring to
> we must make a copy of both arrays (templateIds and designatedWallets) before calling registerDapi() function.

Reply:
Sure, I don't think I was clear enough on this point. Since we are not actually changing the current dapi templateIds but in fact what we need to do is register a new dapi based on the values of the current dapi. This means we need to make a copy of the templateIds and designatedWallets arrays in order to call `registerDapi()`.

For the `addTemplate() `function I was not able to create a copy of the arrays (plus the new id) without using a for loop. Basically because memory arrays in solidty cannot be resized anymore.

I also thought about using [this](https://solidity-by-example.org/app/iterable-mapping/) IterableMapping library and even tough it is pretty neat I still couldn't get rid of all the for loops.

Like you said this is a very common problem in solidity (need to access an element by key and also iterate over the list) but what makes this specific case somewhat special is that we need to make copies of the arrays and not actually modify the arrays stored in state.

Reply:
An alternative way to implement this is that the caller gives the new template ID array, the function checks that the new template ID array is sufficiently similar to the old one, then the new array gets used to create the new dAPI